File: src/main/java/com/burra/practice/leetcode/GroupAnagrams.java
package com.burra.practice.leetcode;

import java.util.*;

/**
 * Input: ["eat", "tea", "tan", "ate", "nat", "bat"], Output: [ ["ate","eat","tea"], ["nat","tan"], ["bat"] ]
 */
public class GroupAnagrams {

  public List<List<String>> groupAnagrams(String[] strs) {
    Map<String, List<String>> map = new HashMap<>();
    for (String s : strs) {
      String count = chartCountToString(s);
      if (!map.containsKey(count)) {
        map.put(count, new ArrayList<String>());
      }
      map.get(count).add(s);

    }
    return new ArrayList(map.values());
  }

  public String chartCountToString(String s) {
    int[] nums = new int[26];
    Arrays.fill(nums, 0);
    for (char c : s.toCharArray()) {
      nums[c - 'a']++;
    }
    StringBuilder sb = new StringBuilder("");
    for (int i = 0; i < nums.length; i++) {
      sb.append('#');
      sb.append(nums[i]);
    }
    return sb.toString();
  }
}

File: src/main/java/com/burra/practice/blind75/GroupAnagrams.kt
package com.burra.practice.blind75

/**
 * Given an array of strings strs, group the anagrams together. You can return the answer in any order.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all
the original letters exactly once.



Example 1:

Input: strs = ["eat","tea","tan","ate","nat","bat"]
Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
Example 2:

Input: strs = [""]
Output: [[""]]
Example 3:

Input: strs = ["a"]
Output: [["a"]]


Constraints:

1 <= strs.length <= 104
0 <= strs[i].length <= 100
strs[i] consists of lowercase English letters.
 */
class GroupAnagrams {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val map = hashMapOf<String, MutableList<String>>()

        strs.forEach {
            val countsString = countOfCharacter(it)
            val temp = map.getOrDefault(countsString, mutableListOf())
            temp.add(it)
            map[countsString] = temp
        }

        return map.values.toList()
    }

    private fun countOfCharacter(it: String): String {
        val array = IntArray(26)
        it.forEach { array[it - 'a']++ }
        val sb = StringBuilder()
        array.forEach {
            sb.append(it)
            sb.append("#")
        }
        return sb.toString()
    }
}